core/src/main/java/kafka/server/share/SharePartition.java

```diff
@@ -371,30 +372,52 @@ public static RecordState forId(byte id) {
     public CompletableFuture<Void> maybeInitialize() {
         log.debug("Maybe initialize share partition: {}-{}", groupId, topicIdPartition);
         CompletableFuture<Void> future = new CompletableFuture<>();
+        AtomicReference<Optional<Throwable>> futureException = new AtomicReference<>(Optional.empty());
         // Check if the share partition is already initialized.
-        maybeCompleteInitialization(future);
-        if (future.isDone()) {
+        InitializationResult initializationResult = maybeCompleteInitialization();
+        if (initializationResult.isComplete()) {
+            if (initializationResult.throwable() != null) {
+                future.completeExceptionally(initializationResult.throwable());
+            } else {
+                future.complete(null);
+            }
             return future;
         }
 
         // All the pending requests should wait to get completed before the share partition is initialized.
         // Attain lock to avoid any concurrent requests to be processed.
         lock.writeLock().lock();
+        boolean shouldFutureBeCompleted = false;
         try {
             // Re-check the state to verify if previous requests has already initialized the share partition.
-            maybeCompleteInitialization(future);
-            if (future.isDone()) {
+            initializationResult = maybeCompleteInitialization();
+            if (initializationResult.isComplete()) {
+                if (initializationResult.throwable() != null) {
+                    futureException.set(Optional.of(initializationResult.throwable()));
+                } else {
+                    futureException.set(Optional.empty());
```

Comment:
That is its initial value, so you don't need to set the exception here.

Reply:
done